mava/evaluator.py

```diff
@@ -12,329 +12,140 @@
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-        )
-        absolute_metric_evaluator = get_rnn_evaluator_fn(
-            eval_env,
-            network_apply_fn,  # type: ignore
-            config,
-            scanned_rnn,
-            log_win_rate,
-            10,
-        )
-    else:
-        evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate  # type: ignore
-        )
-        absolute_metric_evaluator = get_ff_evaluator_fn(
-            eval_env, network_apply_fn, config, log_win_rate, 10  # type: ignore
-        )
-
-    evaluator = jax.pmap(evaluator, axis_name="device")
-    absolute_metric_evaluator = jax.pmap(absolute_metric_evaluator, axis_name="device")
-
-    return evaluator, absolute_metric_evaluator
+
+    def eval_fn(params: FrozenDict, key: PRNGKey, init_act_state: ActorState) -> Dict[str, Array]:
+        def _env_step(eval_state: _EvalEnvStepState, _: Any) -> Tuple[_EvalEnvStepState, TimeStep]:
+            """Performs a single environment step"""
+            env_state, ts, key, actor_state = eval_state
+
+            key, act_key = jax.random.split(key)
+            action, actor_state = act_fn(params, ts, act_key, **actor_state)
+            env_state, ts = jax.vmap(env.step)(env_state, action)
+
+            return (env_state, ts, key, actor_state), ts
+
+        def _episode(key: PRNGKey, _: Any) -> Tuple[PRNGKey, Metrics]:
+            """Simulates `config.arch.num_envs` episodes."""
+            key, reset_key = jax.random.split(key)
+            reset_keys = jax.random.split(reset_key, config.arch.num_envs)
+            env_state, ts = jax.vmap(env.reset)(reset_keys)
+
+            step_state = env_state, ts, key, init_act_state
+            _, timesteps = jax.lax.scan(_env_step, step_state, jnp.arange(env.time_limit))
+
+            metrics = timesteps.extras["episode_metrics"]
+            if config.env.log_win_rate:
+                metrics["won_episode"] = timesteps.extras["won_episode"]
+
+            # find the first instance of done to get the metrics at that timestep we don't
+            # care about subsequent steps because we only the results from the first episode
+            done_idx = jnp.argmax(timesteps.last(), axis=0)
+            metrics = jax.tree_map(lambda m: m[done_idx, jnp.arange(config.arch.num_envs)], metrics)
+            del metrics["is_terminal_step"]  # uneeded for logging
+
+            return key, metrics
+
+        # todo: do we want to divide this by n devices?
+        episode_loops = math.ceil(eval_episodes / config.arch.num_envs)
+        # This loop is important because we don't want too many parallel envs.
+        # So in evaluation we have config.arch.num_envs parallel envs and loop
+        # enough times so that we do at least `eval_episodes` number of episodes
+        _, metrics = jax.lax.scan(_episode, key, xs=None, length=episode_loops)
```

Comment:
We need to consider the case when eval_episodes<num_envs

Reply:
I see two options: 
1. warning and do `num_envs` episodes
2. just do `eval_episdoes` parallel envs once

Reply:
I prefer the second because you know sometimes we do 64 parallel envs but we want to follow evaluation protocol by making only 32 evaluation episodes :pray: